src/sentry/static/sentry/app/components/hovercard.jsx

```diff
@@ -46,62 +54,150 @@ class Hovercard extends React.Component {
-
-    const classes = {
-      'hovercard-bottom': rect.top < 0,
-      'hovercard-left': rect.left < 0,
-      'hovercard-right': rect.right > window.innerWidth && !(rect.left < 0),
-    };
+  positionCard() {
+    if (!this.cardElement.current || this.state.visible) return;
+    const rect = this.cardElement.current.getBoundingClientRect();
 
-    // If it's already been given a positon class do not use the recomputed
-    // position classes, since they will have been computed from it's new
-    // current correct position. Use the previous positions.
-    const currentClasses = intersection(this.cardElement.classList, Object.keys(classes));
+    const y = rect.top < 0 ? 'bottom' : 'top';
+    const x =
+      rect.right > window.innerWidth && !(rect.left < 0)
+        ? 'right'
+        : rect.left < 0 ? 'left' : 'middle';
 
-    return currentClasses.length > 0 ? currentClasses : classes;
+    this.setState({position: {x, y}});
   }
 
   render() {
     const {containerClassName, className, header, body} = this.props;
-    const {rendered, visible} = this.state;
+    const {rendered, visible, position} = this.state;
 
-    const containerCx = classNames('hovercard-container', containerClassName);
-    const cx = classNames('hovercard', this.positionClasses(), className, {
-      'with-header': header,
-      visible,
-    });
+    // Maintain the hovercard class name for BC with less styles
+    const cx = classNames('hovercard', className);
 
     return (
-      <span
-        className={containerCx}
+      <Container
+        className={containerClassName}
         onMouseEnter={this.handleToggleOn}
         onMouseLeave={this.handleToggleOff}
       >
         {this.props.children}
         {rendered && (
-          <div
+          <StyledHovercard
+            visible={visible}
+            withHeader={!!header}
             className={cx}
-            ref={e => {
-              this.cardElement = e;
-            }}
+            position={position || DEFAULT_POSITION}
+            innerRef={this.cardElement}
           >
-            <div className="hovercard-hoverlap" />
-            {header && <div className="hovercard-header">{header}</div>}
-            {body && <div className="hovercard-body">{body}</div>}
-          </div>
+            {header && <Header>{header}</Header>}
+            {body && <Body>{body}</Body>}
+          </StyledHovercard>
         )}
-      </span>
+      </Container>
     );
   }
 }
 
+const translateX = x => `translateX(${x === 'middle' ? '-50%' : 0})`;
+const slideTranslateY = y => `translateY(${{top: -1, bottom: 1}[y] * 14}px)`;
+
+const slideIn = p => keyframes`
+  from {
+    transform: ${translateX(p.position.x)} ${slideTranslateY(p.position.y)};
+  }
+  to {
+    transform: ${translateX(p.position.x)} translateY(0);
+  }
+`;
+
+const positionX = offset => p => css`
+  right: ${p.position.x === 'right' ? offset : 'inherit'};
+  left: ${{left: offset, right: 'inherit', middle: '50%'}[p.position.x]};
```

Comment:
This is difficult to read/understand at first glance, imo.

Reply:
Do you think a ternary would be more clear?

```js
p.position.x === 'left' ? offset : p.position.x === 'middle' ? '50%' : 'inherit'
```

I don't particularly like it as much since it's not as clear that it also handles 'right'.

Reply:
maybe a get function with a case switch?

Reply:
How do you guys feel about this:

```jsx
const positionX = offset => p => css`
  left: ${{
    left: offset,
    right: 'inherit',
    middle: '50%',
  }[p.position.x]};
  right: ${p.position.x === 'right' ? offset : 'inherit'};

  transform: ${translateX(p.position.x)};
`;
```

Reply:
I just think that inline object lookup syntax will not be clear especially since we don't really use it anywhere else?

Reply:
86e44f1 Just switched it to use ternaries.